### PR TITLE
Initialize return variables on C++

### LINKED ISSUE #5278: [BUG] UB from __Pyx_pretend_to_initialize not initializing
### Describe the bug

Apologies in advance: I don't use cython, but I'm debugging a bug caused by using cython with a fresh version of clang. I'll try to get a better repro soon, but for now I'm just filing this in case the cause/fix is obvious. My repro is representative of something internal, but I haven't reduced it to something standalone yet.

With a very new [LLVM commit](https://github.com/llvm/llvm-project/commit/b6a0be8ce3114d0c57e7a7d6c3c222986ca506ad), we're seeing failures in Cython code. The issue seems to be caused by UB in Cython when returning uninitialized data.

Given some Cython code like this:

```cython
T decode(string encoded) except *:
  cdef T result
    if this.decoder.decode(encoded, &result):
      return result
  raise ValueError(f'Failed to decode {repr(encoded)}')

decode(.... <something invalid> ...)
```

The generated code for will look something like this:
```c++
static CYTHON_INLINE void __Pyx_pretend_to_initialize(void* ptr) { (void)ptr; }

template <typename T>
T decode(std::string __pyx_v_encoded) {
  T __pyx_v_result;
  T __pyx_r;
  ...
  __pyx_t_1 = (this->decoder.decode(__pyx_v_encoded, &__pyx_v_result) != 0);
  if (__pyx_t_1) {
    __pyx_r = __pyx_v_result;
    goto __pyx_L0;
  }
  ...
__pyx_L1_error:;
  __Pyx_pretend_to_initialize(&__pyx_r);
__pyx_L0:;
  return __pyx_r;  // This is only *actually* initialized in the happy case.
}
```

The UB is that `__Pyx_pretend_to_initialize` doesn't actually do any initialization, and so `decode` will return the uninitialized `__pyx_r` value. It's enough to silence static compiler warnings, but the optimizer will see that it doesn't do anything to the variable, and mis-optimize the UB accordingly.

There are two possible changes to `__Pyx_pretend_to_initialize` that would make it satisfy this requirement:

```c++
// Prevent the compiler from optimizing this away, e.g. see
// https://stackoverflow.com/questions/28287064/how-not-to-optimize-away-mechanics-of-a-folly-function
static CYTHON_INLINE void __Pyx_pretend_to_initialize(void* ptr) { asm volatile("" : "+r"(ptr)); }

// Or, make it actually do something. The compiler is hopefully smart enough to optimize this away, but I haven't checked.
// Requires changing the calls from __Pyx_pretend_to_initialize(&x) to __Pyx_pretend_to_initialize(&x, sizeof(x))
static CYTHON_INLINE void __Pyx_pretend_to_initialize(void* ptr, size_t size) { memset(ptr, 0, size); }
```

### Code to reproduce the behaviour:

```cython
T decode(string encoded) except *:
  cdef T result
    if this.decoder.decode(encoded, &result):
      return result
  raise ValueError(f'Failed to decode {repr(encoded)}')

decode(.... <something invalid> ...)
```

### Expected behaviour

When passed invalid data, it should raise an error. Instead, it crashes.

### Environment

OS: Linux
Python version: 3.10
Cython version: 0.29.32 (but I still see the bug in trunk)
LLVM/Clang version: something very recent; newer than this: https://github.com/llvm/llvm-project/commit/b6a0be8ce3114d0c57e7a7d6c3c222986ca506ad

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by da-woods**
rather than relying on __Pyx_pretend_to_initialize which isn't sufficient to avoid undefined behaviour there.

Fixes #5278

### CODE REVIEW

File: Cython/Compiler/Nodes.py [hunk 1]
```diff
@@ -2267,6 +2273,17 @@ def assure_gil(code_path, code=code):
                     code.putln("%s = %s;" % (Naming.retval_cname, err_val))
             elif not return_type.is_void:
                 code.putln("__Pyx_pretend_to_initialize(&%s);" % Naming.retval_cname)
+                # For C, __Pyx_pretend_to_initialize() is sufficient (taking an address
+                # is enough to make returning a variable defined). For C++ this isn't
+                # true, and we genuinely have to make sure a variable is initialized.
+                empty_decl_code = return_type.empty_declaration_code()
+                # C++ doesn't want names of the form '= struct foo()' so
+                # remove the struct
+                empty_decl_code = empty_decl_code.replace("struct ", "").replace("union ", "")
```
Comment (scoder) [inline Cython/Compiler/Nodes.py:2282]: I'm trying to think of situations where this simple replacement text might not be safe. What about function pointers, for example? Couldn't those have spaces after user defined type names, like a typedef name `sometypestruct` ending in `…struct ` with a space?

Comment (da-woods) [reply Cython/Compiler/Nodes.py:2282]: I think a function pointer would end to using `NULL` to signal an error so wouldn't get used by the "uninitialised return on error" path.

However, I take the general point here - I'll have another look. Worst case would be to restrict the replacement to the start of strings...

I think I've also missed `enum ` although they might also have a well-defined error return value.

(I think this change can miss the next beta though)

Comment (da-woods) [reply Cython/Compiler/Nodes.py]: I've done this in PyrexTypes instead, which is a bit more verbose but should be more reliable

**Comment by da-woods** [comment]
I'll use the newer attempt instead.

**Pull request closed without merging**